File: sdk/game-access/src/verify.js
import { createPublicKey, verify } from 'node:crypto';
import {
  CLOCK_SKEW_MS, MAX_OBSERVATION_AGE_MS, actionContract, canonical,
  challengeValues, evidenceDigest, hex, identifier, observationMessage, requireCondition, units,
} from './contract.js';

/** Public key is pinned by the trusted caller, never learned from an observation. */
export function verifyObservation({ expectedChallenge, observation, mainIssuer, keyId, publicKeyHex, nowMs }) {
  identifier(mainIssuer);
  identifier(keyId);
  hex(publicKeyHex, 32);
  const now = units(nowMs);
  const expected = canonical(challengeValues(expectedChallenge));
  const actual = canonical(challengeValues(observation.challenge));
  requireCondition(expected.equals(actual), 'challenge_mismatch');
  requireCondition(expectedChallenge.main_issuer === mainIssuer && observation.key_id === keyId, 'authority_mismatch');
  const message = observationMessage(observation);
  const observed = units(observation.observed_at_ms);
  const from = units(observation.grant.not_before_ms);
  const until = units(observation.grant.expires_at_ms);
  requireCondition(now >= from && now < until && observed >= from && observed < until
    && observed <= now + CLOCK_SKEW_MS && now < observed + MAX_OBSERVATION_AGE_MS, 'observation_expired');
  requireCondition(observation.grant.scopes.includes(actionContract(expectedChallenge.action).scope), 'scope_missing');
  const publicKey = createPublicKey({
    key: Buffer.concat([Buffer.from('302a300506032b6570032100', 'hex'), Buffer.from(publicKeyHex, 'hex')]),
    format: 'der', type: 'spki',
  });
  requireCondition(verify(null, message, publicKey, Buffer.from(observation.signature_hex, 'hex')), 'signature_invalid');
  return Object.freeze({
    schema: 'esk.game.session.verified_observation.v1',
    main_user_id: observation.grant.main_user_id,
    main_session_id: observation.grant.main_session_id,
    grant_id: observation.grant.grant_id,
    revision: observation.grant.revision,
    valid_until_ms: (until < observed + MAX_OBSERVATION_AGE_MS ? until : observed + MAX_OBSERVATION_AGE_MS).toString(),
    authorization_digest: evidenceDigest(message, observation.signature_hex),
    // Deliberately not a bearer, principal, wallet binding, or payment permission.
    wallet_bound: false, funds_moved: false,
  });
}

File: server/src/esk_platform/game_access/protocol/verify.rs
use super::{wire, *};
use ed25519_dalek::{Signature, VerifyingKey};

pub fn verify_observation_json(
    input: &[u8],
    expected: &Challenge,
    authority: &Authority,
    now_ms: u64,
) -> Result<VerifiedObservation, Error> {
    if input.len() > 16_384 {
        return Err(Error::InvalidContract);
    }
    let observation = serde_json::from_slice(input).map_err(|_| Error::InvalidContract)?;
    verify_observation(&observation, expected, authority, now_ms)
}

pub fn verify_observation(
    observation: &Observation,
    expected: &Challenge,
    authority: &Authority,
    now_ms: u64,
) -> Result<VerifiedObservation, Error> {
    wire::id(&authority.main_issuer)?;
    wire::id(&authority.key_id)?;
    if now_ms > i64::MAX as u64 {
        return Err(Error::InvalidContract);
    }
    if challenge_digest(&observation.challenge)? != challenge_digest(expected)? {
        return Err(Error::ChallengeMismatch);
    }
    if expected.main_issuer != authority.main_issuer || observation.key_id != authority.key_id {
        return Err(Error::AuthorityMismatch);
    }
    let message = observation_message(observation)?;
    let now = u128::from(now_ms);
    let observed = u128::from(wire::units(&observation.observed_at_ms)?);
    let from = u128::from(wire::units(&observation.grant.not_before_ms)?);
    let until = u128::from(wire::units(&observation.grant.expires_at_ms)?);
    if now < from
        || now >= until
        || observed < from
        || observed >= until
        || observed > now + 2_000
        || now >= observed + 10_000
    {
        return Err(Error::ObservationExpired);
    }
    if !observation
        .grant
        .scopes
        .iter()
        .any(|s| s == expected.action.scope())
    {
        return Err(Error::ScopeMissing);
    }
    let key =
        VerifyingKey::from_bytes(&authority.public_key).map_err(|_| Error::SignatureInvalid)?;
    let signature = Signature::from_slice(&wire::hex(&observation.signature_hex, 64)?)
        .map_err(|_| Error::SignatureInvalid)?;
    key.verify_strict(&message, &signature)
        .map_err(|_| Error::SignatureInvalid)?;
    Ok(VerifiedObservation {
        main_user_id: observation.grant.main_user_id.clone(),
        main_session_id: observation.grant.main_session_id.clone(),
        grant_id: observation.grant.grant_id.clone(),
        revision: observation.grant.revision.clone(),
        valid_until_ms: until.min(observed + 10_000).to_string(),
        authorization_digest: authorization_digest(observation)?,
        wallet_bound: false,
        funds_moved: false,
    })
}
